### PR TITLE
Added limit for size of configuration file that tenant can upload to Alertmanager.

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**: This PR introduces new limit in Alertmanager: limit for size of configuration file that tenant can upload.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/alertmanager/api.go [hunk 1]
```diff
@@ -98,13 +100,30 @@ func (am *MultitenantAlertmanager) SetUserConfig(w http.ResponseWriter, r *http.
...
+	payload, err := ioutil.ReadAll(input)
 	if err != nil {
 		level.Error(logger).Log("msg", errReadingConfiguration, "err", err.Error())
 		http.Error(w, fmt.Sprintf("%s: %s", errReadingConfiguration, err.Error()), http.StatusBadRequest)
 		return
 	}
 
+	if maxConfigSize > 0 && len(payload) > maxConfigSize {
+		msg := fmt.Sprintf(errConfigurationTooBig, maxConfigSize)
+		level.Error(logger).Log("msg", msg)
```
Comment (gotjosh) [inline pkg/alertmanager/api.go:122]:
```suggestion
		level.Error(logger).Log("msg", msg, "user", userID)
```

nit: I noticed none of these logs in this function include the userID, I wouldn't say change them all (but maybe we do 😄 ) but let's start the trend.

**Review by gotjosh — APPROVED** [review]
LGTM 👏  great work.

File: pkg/alertmanager/api.go (hunk 1 above)
Comment (pstibrany) [reply pkg/alertmanager/api.go:122]: The reason why logger doesn’t need user is that the following line at the top of this method:

```
logger := util_log.WithContext(r.Context(), am.logger)
```

will already put tenant into logger, if one is present in the context.

Comment (gotjosh) [reply pkg/alertmanager/api.go:122]: TIL, all good then.

Comment (pstibrany) [reply pkg/alertmanager/api.go:122]: Thanks for your review!

File: pkg/util/validation/limits.go [hunk 2]
```diff
@@ -173,6 +174,7 @@ func (l *Limits) RegisterFlags(f *flag.FlagSet) {
 		l.NotificationRateLimitPerIntegration = NotificationRateLimitMap{}
 	}
 	f.Var(&l.NotificationRateLimitPerIntegration, "alertmanager.notification-rate-limit-per-integration", "Per-integration notification rate limits. Value is a map, where each key is integration name and value is a rate-limit (float). On command line, this map is given in JSON format. Rate limit has the same meaning as -alertmanager.notification-rate-limit, but only applies for specific integration. Allowed integration names: "+strings.Join(allowedIntegrationNames, ", ")+".")
+	f.IntVar(&l.AlertmanagerMaxConfigSize, "alertmanager.max-config-size", 0, "Maximum size of configuration file for Alertmanager that tenant can upload via Alertmanager API. 0 = no limit.")
```
Comment (pracucci) [inline pkg/util/validation/limits.go:177]: [nit] I'm wondering if, to make it even more clear, we should mention it's in bytes. We can do it in the CLI flag description but I'm also wondering if we should add a `-bytes` suffix to the config like we have for other configs.

File: pkg/alertmanager/api.go (hunk 1 above)
Comment (pracucci) [inline pkg/alertmanager/api.go:122]: This is not an actionable error on Cortex side, so I wouldn't use the error level. I would use Warn instead.

**Review by pracucci — APPROVED** [review]
Good job! LGTM (modulo a couple of nits I would be glad if you could take a look at)

File: pkg/util/validation/limits.go (hunk 2 above)
Comment (pstibrany) [reply pkg/util/validation/limits.go:177]: Sounds good. I will update it.

**Pull request merged**